### ISSUE TITLE
region is not picked up from the AWS profile

### ISSUE THREAD

**Issue by ringods** [terraform-providers/terraform-provider-aws#7750]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.11
+ provider.aws v1.60.0
```

#### AWS config files on disk

`~/.aws/credentials`
```ini
[myprofile]
aws_access_key_id=<my access key>
aws_secret_access_key=<my secret key>
```

`~/.aws/config`
```ini
[myprofile]
region=eu-west-1
output=json
```

#### scenario 1: profile set via the provider block.

```hcl
provider "aws" {
  version = "~> 1.60"
  profile = "myprofile"
}
```

#### scenario 2: profile set via the environment variable

```hcl
provider "aws" {
  version = "~> 1.60"
}
```
with
```sh
$ export AWS_PROFILE=myprofile
```

### Expected Behavior

The Argument Reference docs state this about `region`:
> region - (Required) This is the AWS region. It must be provided, but it can also be sourced from the AWS_DEFAULT_REGION environment variables, or via a shared credentials file if profile is specified.

When running `terraform plan`, I expect the `region` to be picked up via the `myprofile` section in `~/.aws/config`.

### Actual Behavior

Terraform asks for the `region` value:

```sh
$ terraform plan
provider.aws.region
  The region where AWS operations will take place. Examples
  are us-east-1, us-west-2, etc.

  Default: us-east-1
  Enter a value:
```

### Steps to Reproduce

1. `terraform init`
1. `terraform plan`

### Arguments

The reason this fails is because `region` is a `Required` attribute for the AWS provider. The code only retrieves the value from the 2 environment variables when not specified in HCL. This is clearly not as documented with `via a shared credentials file if profile is specified.`

https://github.com/terraform-providers/terraform-provider-aws/blob/687f09ebf7d2b5d0fe1b92f45c3c49ead8cb3711/aws/provider.go#L60-L69

The AWS SDK, like it does for credentials, searches for `region` via environment variables and profile when not specified explicitly:

https://github.com/aws/aws-sdk-go/blob/275272fc5c7fdea1719f5851925dc5d9df27f89d/aws/session/env_config.go#L205

https://github.com/aws/aws-sdk-go/blob/275272fc5c7fdea1719f5851925dc5d9df27f89d/aws/session/shared_config.go#L236-L239

It probably needs a bit more testing, but I would suggest making `region` an `Optional` attribute and remove the `DefaultFunc` initialization on the AWS provider side.

**Comment by grimm26** [comment]
I believe that this is just how the AWS SDK works.  I run into this when writing my own tools using the SDK, and I have to dig into Aws.shared_config to pull out the region from the profile config and inject that into Aws.config.  
In terraform, I always specify a region when instantiating a provider as that explicitly locks down the config to act on the correct region.  Otherwise if someone happens to have their profile set to another region, bad things will happen :)

**Comment by mattburgess** [comment]
@ringods not sure if it's pertinent to your issue, but your `~/.aws/config` file doesn't look right to me.  As per the docs at https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html profiles in the config file should be started with `[profile profilename]` which is confusingly different to the `[profilename]` format used in the credentials file.  So your example should be:

```
[profile myprofile]
region=eu-west-1
output=json
```

**Comment by ringods** [comment]
@mattburgess good catch, but even after fixing my `~/.aws/config`, this bug still stands unfortunately.

**Comment by scalp42** [comment]
Can confirm here, always asking for region.

**Comment by nywilken** [comment]
Related #8256

**Comment by deitch** [comment]
Has anyone come up with any decent workarounds for this? I don't love messy HCL or other "temporary fixes", but happy to do what I need just to get this to work?

**Comment by chrispruitt** [comment]
Any update on this?

I am seeing this in terraform v0.12.20

My shared config: (just using the default profile)
```
[default]
aws_access_key_id = MYKEY
aws_secret_access_key = MYSECRET
region = us-east-1
```

when I run `terraform plan` i get:
```
Error: Missing required argument

The argument "region" is required, but was not set.
```

Everything works if I explicitly export the `AWS_DEFAULT_REGION` var.
My credentials are being picked up from the file properly, but not the region for some reason.

Any help would be appreciated!

**Comment by t0rr3sp3dr0** [comment]
This is still present on v2.61.0 of the provider

**Comment by peterchaula** [comment]
Mine worked after adding the region to the credentials file  as in @chrispruitt's comment

**Comment by MohammedWahid** [comment]
still present

**Comment by cmeisinger** [comment]
Confirming this is still an issue in v0.15.3.

**Comment by Jonas-Noh** [comment]
Confirming this is still an issue in v0.15.5.

**Comment by mparkerdurost-jama** [comment]
This is still an issue in v4.48.0 of the provider.

**Comment by n2N8Z** [comment]
Still broken in v4.55.0.

**Comment by imranismail** [comment]
For those looking for a workaround, I did it this way:

```terraform
data "config_ini" "aws_config" {
  ini = file("~/.aws/config")
  section = "profile ${terraform.workspace}"
}

locals {
  aws_config = {for key, val in jsondecode(data.config_ini.aws_config.json) : trimspace(key) => trimspace(val)}
}

provider "aws" {
  profile  = terraform.workspace
  region  = local.aws_config.region
}
```

**Comment by beck3905** [comment]
I know this is an old issue, but I just came across it. I have `region` with a literal string in my provider config in main.tf. When I run `tofu apply` I'm still getting prompted that I must set the `AWS_REGION` or `AWS_DEFAULT_REGION` environment variable. I also have the region specified in my AWS profile and I've set the `AWS_PROFILE` environment variable.
